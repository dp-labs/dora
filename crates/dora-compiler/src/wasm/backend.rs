use dora_ir::IRTypes;
use dora_runtime::symbols;
use melior::dialect::{
    arith::{self, CmpfPredicate, CmpiPredicate},
    cf, func,
};
use melior::ir::{
    attribute::FlatSymbolRefAttribute, r#type::IntegerType, BlockRef, OperationRef, Region, Type,
    TypeLike, Value, ValueLike,
};
use wasmer_types::TrapCode;

use crate::backend::IntCC;
use crate::errors::Result;
use crate::value::ToContextValue;
use crate::{backend::TypeMethods, context::Context, conversion::builder::OpBuilder, state::State};

use super::intrinsics::WASMIntrinsics;

/// A builder structure for constructing WebAssembly (WASM) operations and managing
/// the underlying compilation process using an `OpBuilder`. The `WASMBuilder`
/// operates on a `WASMBackend` and provides an interface to emit operations
/// within the context of a given WASM execution model.
///
/// # Fields:
/// - `backend`: A mutable reference to the `WASMBackend`, which contains the overall
///   context and state of the WASM environment being built.
/// - `builder`: An `OpBuilder` instance responsible for generating operations
///   during the construction process.
///
/// # Example Usage:
/// ```no_check
/// let mut wasm_builder = WASMBuilder {
///     backend: &mut wasm_backend,
///     builder: op_builder,
/// };
/// wasm_builder.builder.some_build_method();
/// ```
///
/// # Notes:
/// - The `WASMBuilder` works closely with the `WASMBackend` to manage WebAssembly-specific
///   context and state, allowing for the construction of WASM operations within the MLIR system.
#[derive(Debug)]
pub struct WASMBuilder<'c, 'a> {
    /// A mutable reference to the WASM backend containing the context and state.
    pub backend: &'a mut WASMBackend<'c>,
    /// The builder used to generate operations within the WASM execution context.
    pub builder: OpBuilder<'c, 'a>,
}

impl<'c> std::ops::Deref for WASMBuilder<'c, '_> {
    type Target = WASMBackend<'c>;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.backend
    }
}

impl std::ops::DerefMut for WASMBuilder<'_, '_> {
    #[inline]
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.backend
    }
}

/// The backend structure for managing the WebAssembly (WASM) execution context and
/// intrinsics. `WASMBackend` encapsulates the relevant MLIR context, intrinsic types,
/// and execution state required for compiling and executing WASM bytecode.
///
/// # Fields:
/// - `ctx`: A reference to the MLIR `Context`, which provides the underlying infrastructure
///   for managing the compilation and execution process.
/// - `intrinsics`: A set of WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64,
///   that are commonly used in WASM operations.
/// - `state`: The current state of the execution, which includes the stack and control
///   structures required for managing the flow of WASM bytecode execution.
///
/// # Example Usage:
/// ```no_check
/// let wasm_backend = WASMBackend {
///     ctx: &mlir_context,
///     intrinsics: wasm_intrinsics,
///     state: wasm_state,
/// };
/// ```
///
/// # Notes:
/// - The `WASMBackend` is responsible for maintaining the state and context necessary
///   for WASM execution and compilation. It interacts with other structures like the
///   `WASMBuilder` to facilitate the construction of operations that adhere to the
///   WebAssembly execution model.
#[derive(Debug)]
pub struct WASMBackend<'c> {
    /// A reference to the MLIR context used in the WASM backend.
    pub ctx: &'c Context,
    /// WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64.
    pub intrinsics: WASMIntrinsics<'c>,
    /// The current state of WASM execution, including the stack and control structures.
    pub state: State<'c, 'c>,
}

impl<'c> WASMBackend<'c> {
    pub fn new(ctx: &'c Context) -> Self {
        Self {
            ctx,
            intrinsics: WASMIntrinsics::declare(ctx),
            state: State::new(),
        }
    }
}

impl<'c> IRTypes for WASMBackend<'c> {
    type Type = Type<'c>;
    type Value = Value<'c, 'c>;
    type Region = Region<'c>;
    type BasicBlock = BlockRef<'c, 'c>;
    type Operation = OperationRef<'c, 'c>;
}

impl TypeMethods for WASMBackend<'_> {
    fn type_ptr(&self) -> Self::Type {
        self.intrinsics.ptr_ty
    }

    fn type_int(&self, bits: u32) -> Self::Type {
        IntegerType::new(&self.ctx.mlir_context, bits).into()
    }
}

pub(crate) fn is_zero<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    value: Value<'c, 'a>,
) -> Result<Value<'c, 'a>> {
    let ty = value.r#type();
    if ty.is_integer() {
        Ok(builder
            .create(arith::cmpi(
                builder.context(),
                CmpiPredicate::Eq,
                value,
                if builder.int_ty_width(ty)? == 32 {
                    builder.create(builder.iconst_32(0)).result(0)?.into()
                } else {
                    builder.create(builder.iconst_64(0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else if ty.is_float() {
        Ok(builder
            .create(arith::cmpf(
                builder.context(),
                CmpfPredicate::Oeq,
                value,
                if ty.is_f32() {
                    builder.create(builder.fconst_32(0.0)).result(0)?.into()
                } else {
                    builder.create(builder.fconst_64(0.0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else {
        // WASM pointer type
        let value = builder.make(builder.load(value, builder.i32_ty()))?;
        let result = builder.make(builder.icmp_imm(IntCC::Equal, value, 0)?)?;
        Ok(unsafe { Value::from_raw(result.to_raw()) })
    }
}

pub(crate) fn trap<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    code: TrapCode,
    continue_block: BlockRef<'c, 'a>,
) -> Result<()> {
    let ctx = builder.ctx;
    let code = builder.make(builder.iconst_32(code as _))?;
    builder.create(func::call(
        ctx,
        FlatSymbolRefAttribute::new(ctx, symbols::wasm::RAISE_TRAP),
        &[code],
        &[],
        builder.get_insert_location(),
    ));
    builder.create(cf::br(&continue_block, &[], builder.get_insert_location()));
    Ok(())
}
