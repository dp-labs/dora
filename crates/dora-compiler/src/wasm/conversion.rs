use crate::conversion::rewriter::{replace_op, Replacer, Rewriter};
use crate::conversion::walker::walk_operation;
use crate::errors::Result;
use crate::value::IntoContextOperation;
use dora_ir;
use melior::ir::ValueLike;
use melior::{
    dialect::DialectHandle,
    ir::{r#type::TypeId, OperationRef},
    pass::{create_external, ExternalPass, Pass, RunExternalPass},
    Context, ContextRef,
};
use tracing::debug;

#[repr(align(8))]
struct PassId;
static CONVERSION_PASS: PassId = PassId;

/// The `ConversionPass` struct is responsible for transforming specific WASM
/// operations into their Dora counterparts. It walks through the MLIR operation
/// tree, identifies certain WASM instructions, and replaces them with Dora operations.
///
/// This transformation is essential in a system where WebAssembly or other intermediate representations
/// are being processed with additional virtual machine layers, providing flexibility and extensibility.
///
/// # Struct Details:
/// The `ConversionPass` struct is typically used as part of a larger pass framework where WASM bytecode
/// or MLIR IR is being converted into an intermediate representation supported by Dora. This pass
/// focuses on replacing arithmetic and logical instructions.
///
/// ## WASM to Dora Operation Transformations like:
/// - `wasm.add` → `dora.add`
/// - `wasm.sub` → `dora.sub`
/// - `wasm.mul` → `dora.mul`
/// - `wasm.div_s` → `dora.sdiv`
/// - `wasm.div_u` → `dora.div`
///
/// These transformations optimize or translate WASM-specific instructions into a format compatible
/// with Dora or other runtime environments.
///
/// # Errors:
/// This conversion pass returns a `Result<()>`, with possible errors occurring during operand retrieval
/// or operation replacement.
///
/// # Methods:
/// - `run(&mut self, operation: OperationRef<'_, '_>) -> Result<()>`
///   - Walks through the operation tree and applies transformations to recognized WASM operators.
#[derive(Clone, Debug)]
pub struct ConversionPass<'c> {
    /// A reference to the MLIR context, which manages global state and resources required for MLIR operations.
    pub ctx: &'c Context,
}

impl ConversionPass<'_> {
    /// Executes the conversion pass on the provided operation.
    ///
    /// This function transforms specific WASM operators into equivalent
    /// Dora operations. It walks through the operation, checks for
    /// recognized WASM instructions (such as `wasm.add`, `wasm.sub`, etc.), and replaces them with their
    /// corresponding Dora counterparts (like `dora.add`, `dora.sub`, etc.).
    ///
    /// # Parameters:
    /// - `operation`: The root `OperationRef` representing the operation to be transformed.
    ///
    /// # Errors:
    /// This function returns an error if there are issues in retrieving operands or replacing
    /// the operation with the new Dora operations.
    pub(crate) fn run(&mut self, operation: OperationRef<'_, '_>) -> Result<()> {
        let mut wasm_ops = vec![];
        walk_operation(
            operation,
            Box::new(|op| {
                let name = op.name().as_string_ref().as_str().unwrap().to_string();
                if name.starts_with("wasm") {
                    wasm_ops.push(op.to_ctx_operation_ref());
                }
                Ok(())
            }),
        )?;

        for op in wasm_ops {
            let name = op.name().as_string_ref().as_str().unwrap().to_string();
            if name == dora_ir::wasm::AddOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::add(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::SubOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::sub(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::MulOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::mul(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::DivSOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::sdiv(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::DivUOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::div(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::DivSOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::sdiv(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::RemUOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::r#mod(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::RemSOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::smod(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::AndOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::and(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::OrOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::or(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::XOrOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::xor(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::ShlOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::shl(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::ShrSOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::sar(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::ShrUOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::shr(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::RotlOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::rotl(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::RotrOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::rotr(self.ctx, lhs.r#type(), lhs, rhs, op.location()).into(),
                );
            } else if name == dora_ir::wasm::ClzOperation::name() {
                let value = op.operand(0)?;
                replace_op(
                    op,
                    dora_ir::dora::clz(self.ctx, value.r#type(), value, op.location()).into(),
                );
            } else if name == dora_ir::wasm::CtzOperation::name() {
                let value = op.operand(0)?;
                replace_op(
                    op,
                    dora_ir::dora::ctz(self.ctx, value.r#type(), value, op.location()).into(),
                );
            } else if name == dora_ir::wasm::PopcntOperation::name() {
                let value = op.operand(0)?;
                replace_op(
                    op,
                    dora_ir::dora::popcnt(self.ctx, value.r#type(), value, op.location()).into(),
                );
            } else if name == dora_ir::wasm::I32Extend8SOperation::name()
                || name == dora_ir::wasm::I32Extend16SOperation::name()
                || name == dora_ir::wasm::I64Extend8SOperation::name()
                || name == dora_ir::wasm::I64Extend16SOperation::name()
                || name == dora_ir::wasm::I64Extend32SOperation::name()
            {
                let value = op.operand(0)?;
                let rewriter = Rewriter::new_with_op(self.ctx, op);
                let byte = rewriter.make(rewriter.iconst(
                    value.r#type(),
                    rewriter.int_ty_width(value.r#type())? as i64 - 1,
                ))?;
                rewriter.replace_op(
                    op,
                    dora_ir::dora::signextend(self.ctx, value.r#type(), byte, value, op.location())
                        .into(),
                )?;
            } else if name == dora_ir::wasm::SelectOperation::name() {
                let lhs = op.operand(0)?;
                let rhs = op.operand(1)?;
                let cond = op.operand(2)?;
                debug_assert!(lhs.r#type() == rhs.r#type());
                replace_op(
                    op,
                    dora_ir::dora::select(
                        self.ctx,
                        op.result(0)?.r#type(),
                        lhs,
                        rhs,
                        cond,
                        op.location(),
                    )
                    .into(),
                );
            }
        }
        Ok(())
    }
}

impl<'c> RunExternalPass<'c> for ConversionPass<'c> {
    fn construct(&mut self) {}

    fn destruct(&mut self) {}

    fn initialize(&mut self, _context: ContextRef<'c>) {}

    fn run(&mut self, operation: OperationRef<'c, '_>, pass: ExternalPass<'_>) {
        if let Err(err) = self.run(operation) {
            debug!("run wasm conversion pass error: {}", err);
            pass.signal_failure();
        }
    }
}

impl ConversionPass<'_> {
    pub fn into_pass(self) -> Pass {
        create_external(
            self,
            TypeId::create(&CONVERSION_PASS),
            "WASM conversion pass",
            "WASM conversion argument",
            "a WASM conversion pass",
            "",
            &[DialectHandle::func()],
        )
    }
}
