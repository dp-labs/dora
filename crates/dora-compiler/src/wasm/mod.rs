//! Reference: https://github.com/wasmerio/wasmer/tree/main/lib/compiler-llvm
#![allow(dead_code)]

mod backend;
mod code;
mod errors;
mod func;
mod intrinsics;
mod state;
mod ty;

#[cfg(test)]
mod tests;

use dora_primitives::config::OptimizationLevel;
use func::FuncTranslator;
use melior::ir::operation::OperationBuilder;
use melior::ir::{Block, Region};
use melior::ir::{Location, Module as MLIRModule};
use std::sync::Arc;
use wasmer_compiler::{
    FunctionBodyData, ModuleEnvironment, ModuleMiddleware, ModuleTranslationState,
};
use wasmer_types::entity::{EntityRef, PrimaryMap};
use wasmer_types::{
    CompileModuleInfo, Features, FunctionIndex, LocalFunctionIndex, SectionIndex, SignatureIndex,
    Symbol, SymbolRegistry,
};

use crate::errors::CompileError;
use crate::{Context, Module};

/// Represents a WebAssembly (Wasm) to LLVM/MLIR compiler. The `WASMCompiler` struct is responsible for
/// compiling WebAssembly code into LLVM/MLIR, leveraging the configuration provided at the time of
/// instantiation.
///
/// # Fields:
/// - `config`: The configuration settings for the compiler, encapsulated in the `Config` struct.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: false,
///     opt_level: OptimizationLevel::O2,
///     middlewares: vec![],
/// };
/// let wasm_compiler = WASMCompiler::new(config);
///
/// // Access the compiler's configuration
/// let compiler_config = wasm_compiler.config();
/// ```
///
/// # Notes:
/// - The `WASMCompiler` struct encapsulates all compiler functionality for transforming WebAssembly bytecode
///   into a more optimized and executable format using LLVM/MLIR.
/// - The compiler that compiles a WebAssembly module with LLVM/MLIR, translating the Wasm to MLIR IR, optimizing
///   it and then translating to assembly or executing it with JIT mode.
#[derive(Debug, Clone)]
pub struct WASMCompiler<'c> {
    /// The MLIR context used for generating operations and managing their lifetime. It encapsulates the state
    /// of the MLIR infrastructure, including types, modules, and operations. This context is tied to the
    /// lifetime `'c` of the EVMCompiler.
    pub ctx: &'c Context,
    /// The configuration settings for the Wasm compiler.
    pub config: Config,
}

impl<'c> WASMCompiler<'c> {
    /// Creates a new instance of the `WASMCompiler` with the given configuration.
    ///
    /// # Arguments:
    /// - `config`: The `Config` object that defines the compilation settings, including optimization levels
    ///   and any middleware.
    ///
    /// # Returns:
    /// - A `WASMCompiler` instance initialized with the provided configuration.
    #[inline]
    pub fn new(ctx: &'c Context, config: Config) -> WASMCompiler {
        WASMCompiler { ctx, config }
    }

    /// Retrieves the configuration for this `WASMCompiler`.
    ///
    /// # Returns:
    /// - A reference to the `Config` object that was used to configure the compiler.
    ///
    /// # Example:
    /// ```no_check
    /// let compiler_config = wasm_compiler.config();
    /// ```
    #[inline]
    fn config(&self) -> &Config {
        &self.config
    }

    /// Compile the WASM bytes using LLVM/MLIR.
    pub fn compile(&self, data: &[u8]) -> Result<Module<'c>, CompileError> {
        let environ = ModuleEnvironment::new();
        let translation = environ
            .translate(data)
            .map_err(|err| CompileError::Codegen(err.to_string()))?;
        let module = translation.module;
        let compile_info = CompileModuleInfo {
            module: Arc::new(module),
            features: Features::default(),
            memory_styles: PrimaryMap::new(),
            table_styles: PrimaryMap::new(),
        };
        self.compile_module(
            &compile_info,
            translation.module_translation_state.as_ref().unwrap(),
            translation.function_body_inputs,
        )
    }

    fn compile_module(
        &self,
        compile_info: &CompileModuleInfo,
        module_translation: &ModuleTranslationState,
        function_body_inputs: PrimaryMap<LocalFunctionIndex, FunctionBodyData<'_>>,
    ) -> Result<Module<'c>, CompileError> {
        // WASM Module information
        let memory_styles = &compile_info.memory_styles;
        let table_styles = &compile_info.table_styles;
        let wasm_module = &compile_info.module;
        let function_body_inputs = function_body_inputs
            .iter()
            .collect::<Vec<(LocalFunctionIndex, &FunctionBodyData<'_>)>>();
        let functions: Vec<_> = function_body_inputs
            .iter()
            .map(|(i, input)| {
                FuncTranslator::translate(
                    self.ctx,
                    &self.config,
                    wasm_module,
                    module_translation,
                    i,
                    input,
                    memory_styles,
                    table_styles,
                    &ShortNames {},
                )
            })
            .collect();
        // MLIR Module region and block
        let module_block = Block::new(&[]);
        for func in functions {
            module_block
                .append_operation(func.map_err(|err| CompileError::Codegen(err.to_string()))?);
        }
        let module_region = Region::new();
        module_region.append_block(module_block);
        // Build main module
        let op = OperationBuilder::new("builtin.module", Location::unknown(&self.ctx.mlir_context))
            .add_regions([module_region])
            .build()?;
        Ok(Module::new(
            MLIRModule::from_operation(op).expect("module failed to create"),
        ))
    }
}

struct ShortNames {}

impl SymbolRegistry for ShortNames {
    fn symbol_to_name(&self, symbol: Symbol) -> String {
        match symbol {
            Symbol::Metadata => "M".to_string(),
            Symbol::LocalFunction(index) => format!("f{}", index.index()),
            Symbol::Section(index) => format!("s{}", index.index()),
            Symbol::FunctionCallTrampoline(index) => format!("t{}", index.index()),
            Symbol::DynamicFunctionTrampoline(index) => format!("d{}", index.index()),
        }
    }

    fn name_to_symbol(&self, name: &str) -> Option<Symbol> {
        if name.len() < 2 {
            return None;
        }
        let (ty, idx) = name.split_at(1);
        if ty.starts_with('M') {
            return Some(Symbol::Metadata);
        }

        let idx = idx.parse::<u32>().ok()?;
        match ty.chars().next().unwrap() {
            'f' => Some(Symbol::LocalFunction(LocalFunctionIndex::from_u32(idx))),
            's' => Some(Symbol::Section(SectionIndex::from_u32(idx))),
            't' => Some(Symbol::FunctionCallTrampoline(SignatureIndex::from_u32(
                idx,
            ))),
            'd' => Some(Symbol::DynamicFunctionTrampoline(FunctionIndex::from_u32(
                idx,
            ))),
            _ => None,
        }
    }
}

/// Represents the configuration options for the `WASMCompiler`. The `Config` struct defines various
/// settings that influence the compilation process, such as optimization levels and verification options.
///
/// # Fields:
/// - `enable_nan_canonicalization`: A boolean flag indicating whether to enable NaN canonicalization, which
///   standardizes NaN representations in floating-point operations.
/// - `enable_verifier`: A boolean flag to enable or disable the verification phase during compilation, ensuring
///   that the compiled code meets correctness constraints.
/// - `opt_level`: The optimization level to be applied during compilation, which can affect the performance and
///   size of the compiled output.
/// - `middlewares`: A vector of middleware components that modify or extend the behavior of the compilation
///   process. Each middleware is represented as a trait object implementing the `ModuleMiddleware` trait.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: true,
///     opt_level: OptimizationLevel::O3,
///     middlewares: vec![],
/// };
/// ```
///
/// # Notes:
/// - The `Config` struct is essential for controlling the behavior of the `WASMCompiler`. You can tune the
///   compilation process by adjusting these fields based on the requirements of your WebAssembly module.
#[derive(Debug, Clone, Default)]
pub struct Config {
    /// A flag to enable NaN canonicalization during floating-point operations.
    pub(crate) enable_nan_canonicalization: bool,
    /// A flag to enable or disable the verifier during compilation.
    pub(crate) enable_verifier: bool,
    /// The optimization level for the compilation process.
    pub(crate) opt_level: OptimizationLevel,
    /// A collection of middleware components that modify the behavior of the compiler.
    pub(crate) middlewares: Vec<Arc<dyn ModuleMiddleware>>,
}
