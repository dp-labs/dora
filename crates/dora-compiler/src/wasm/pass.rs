use super::conversion;
use crate::errors::Result;
use conversion::ConversionPass;
use melior::{ir::Module as MLIRModule, Context};

/// Define a public function `run` to execute the conversion process.
///
/// # Parameters
/// - `ctx`: An MLIR context, used to manage the global state of MLIR.
/// - `module`: A mutable MLIR module representing the MLIR code to be transformed.
///
/// # Returns
/// - `Result<()>`, indicating success or failure of the operation.
pub fn run(ctx: &Context, module: &mut MLIRModule) -> Result<()> {
    let mut conversion_pass = ConversionPass { ctx };
    conversion_pass.run(module.as_operation())
}
