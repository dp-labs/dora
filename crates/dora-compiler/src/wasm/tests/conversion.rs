//！Reference: https://github.com/WebAssembly/spec/tree/main/test/core
use crate::context::Context;
use crate::wasm::{Config, WASMCompiler};
use wasmer::wat2wasm;

macro_rules! assert_snapshot {
    ($code:expr) => {
        let context = Context::new();
        let compiler = WASMCompiler::new(&context, Config::default());
        let wasm_bytes = wat2wasm($code.as_bytes()).unwrap();
        // Compile WASM Bytecode to MLIR EVM Dialect
        let mut module = compiler
            .compile(&wasm_bytes)
            .expect("failed to compile program");
        // Lowering the WASM dialect to the Dora dialect.
        crate::wasm::pass::run(&context.mlir_context, &mut module.mlir_module)
            .expect("failed to run conversion pass");
        // Lowering the Dora dialect to MLIR builtin dialects.
        crate::dora::pass::run(
            &context.mlir_context,
            &mut module.mlir_module,
            &crate::dora::pass::PassOptions {
                code_size: wasm_bytes.len() as u32,
                ..Default::default()
            },
        )
        .unwrap();
        let op = module.module().as_operation();
        assert!(op.verify());
        insta::assert_snapshot!(op);
    };
}

#[test]
fn sum() {
    assert_snapshot!(
        r#"
(module
  (type $sum_t (func (param i32 i32) (result i32)))
  (func $sum_f (type $sum_t) (param $x i32) (param $y i32) (result i32)
    local.get $x
    local.get $y
    i32.add)
  (export "sum" (func $sum_f)))
"#
    );
}

#[test]
fn and_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "and") (param $x i32) (param $y i32) (result i32) (i32.and (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn and_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "and") (param $x i64) (param $y i64) (result i64) (i64.and (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn or_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "or") (param $x i32) (param $y i32) (result i32) (i32.or (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn or_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "or") (param $x i64) (param $y i64) (result i64) (i64.or (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn xor_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "xor") (param $x i32) (param $y i32) (result i32) (i32.xor (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn xor_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "xor") (param $x i64) (param $y i64) (result i64) (i64.xor (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shl_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shl") (param $x i32) (param $y i32) (result i32) (i32.shl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shl_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shl") (param $x i64) (param $y i64) (result i64) (i64.shl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_s_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_s") (param $x i32) (param $y i32) (result i32) (i32.shr_s (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_s_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_s") (param $x i64) (param $y i64) (result i64) (i64.shr_s (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_u_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_u") (param $x i32) (param $y i32) (result i32) (i32.shr_u (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_u_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_u") (param $x i64) (param $y i64) (result i64) (i64.shr_u (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn select() {
    assert_snapshot!(
        r#"
(module
    (func $select (param $cond i32) (param $a i32) (param $b i32) (result i32)
        local.get $cond
        local.get $a
        local.get $b
        select
    )
    (export "select" (func $select))
)
"#
    );
}
