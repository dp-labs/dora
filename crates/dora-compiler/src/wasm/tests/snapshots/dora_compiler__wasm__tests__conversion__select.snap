---
source: crates/dora-compiler/src/wasm/tests/conversion.rs
expression: op
snapshot_kind: text
---
module {
  func.func @select(%arg0: !llvm.ptr, %arg1: i32, %arg2: i32, %arg3: i32) -> i32 {
    %c1_i64 = arith.constant 1 : i64
    %0 = llvm.alloca %c1_i64 x i32 : (i64) -> !llvm.ptr
    llvm.store %arg1, %0 : i32, !llvm.ptr
    %c1_i64_0 = arith.constant 1 : i64
    %1 = llvm.alloca %c1_i64_0 x i32 : (i64) -> !llvm.ptr
    llvm.store %arg2, %1 : i32, !llvm.ptr
    %c1_i64_1 = arith.constant 1 : i64
    %2 = llvm.alloca %c1_i64_1 x i32 : (i64) -> !llvm.ptr
    llvm.store %arg3, %2 : i32, !llvm.ptr
    %3 = llvm.load %0 : !llvm.ptr -> i32
    %4 = llvm.load %1 : !llvm.ptr -> i32
    %5 = llvm.load %2 : !llvm.ptr -> i32
    %c0_i32 = arith.constant 0 : i32
    %6 = arith.cmpi ne, %5, %c0_i32 : i32
    %7 = arith.select %6, %3, %4 : i32
    cf.br ^bb1(%7 : i32)
  ^bb1(%8: i32):  // pred: ^bb0
    return %8 : i32
  }
}
